From patchwork Tue Sep 15 21:29:40 2020
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Tom Lendacky <thomas.lendacky@amd.com>
X-Patchwork-Id: 11777883
Return-Path: <SRS0=V1hm=CY=vger.kernel.org=kvm-owner@kernel.org>
Received: from mail.kernel.org (pdx-korg-mail-1.web.codeaurora.org
 [172.30.200.123])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id 3678B6CA
	for <patchwork-kvm@patchwork.kernel.org>;
 Tue, 15 Sep 2020 21:45:04 +0000 (UTC)
Received: from vger.kernel.org (vger.kernel.org [23.128.96.18])
	by mail.kernel.org (Postfix) with ESMTP id 0D4EC2078E
	for <patchwork-kvm@patchwork.kernel.org>;
 Tue, 15 Sep 2020 21:45:04 +0000 (UTC)
Authentication-Results: mail.kernel.org;
	dkim=pass (1024-bit key) header.d=amdcloud.onmicrosoft.com
 header.i=@amdcloud.onmicrosoft.com header.b="0exTJDH5"
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1728071AbgIOVo5 (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Tue, 15 Sep 2020 17:44:57 -0400
Received: from mail-dm6nam10on2059.outbound.protection.outlook.com
 ([40.107.93.59]:21344
        "EHLO NAM10-DM6-obe.outbound.protection.outlook.com"
        rhost-flags-OK-OK-OK-FAIL) by vger.kernel.org with ESMTP
        id S1727999AbgIOVas (ORCPT <rfc822;kvm@vger.kernel.org>);
        Tue, 15 Sep 2020 17:30:48 -0400
ARC-Seal: i=1; a=rsa-sha256; s=arcselector9901; d=microsoft.com; cv=none;
 b=aNjYlSLEaXVAWq7mHbyJR92EYSjvUnLlW2/isflj+vwY8Mg5q3bxOwrf8nxSnF5BhfD3zK3y+671ZBsyzEdP28aRSByF2TP5SwBX/6m4i5uVXfWWRVlfKw7jEBoBL3IbGuyVixrlI5AGSJpx+qHRgl1y1kZIHl4OTII9Mh87D2WL6IdRzf1cdTuLgpfotErpjuTBpiOcdE+lONycrNB571a1cet56l7L0l3oI/oBJoh2FiKqp8R1w6SOPP8Ce8PaxEJbJFlyqF6UpDMaL5ZXxERF/FuBmljDq7axfXjqcMBIq70/GEVvdvWBBkIkx/hd38BwFr11ZYIHLvJ7GvQchg==
ARC-Message-Signature: i=1; a=rsa-sha256; c=relaxed/relaxed; d=microsoft.com;
 s=arcselector9901;
 h=From:Date:Subject:Message-ID:Content-Type:MIME-Version:X-MS-Exchange-SenderADCheck;
 bh=MJ+BpJng5csN/LA+I/qcPQkYM/tISIORmRx+LOYMnks=;
 b=bfwY3oY3hj8nfswS/j2xKnUzgMi7NPWmwJbieQaGJiMcze+lTPgCKqjIMowIR6Ib1aCrBpZ11E7PDPEMsJ7EyRIeFg25r4LQBRszH/oQIUK6b1y4Galh9HVQzhwUK5pWwgTnmxF7T3WlT7RW/jjj1QI8GM3GY8ICsGZyUVPnv4tCqCIKqq7lanq8D1d5Wg6nI31wVjAhccjzvDmVuWw098n4iJjQ//7cSPwmM+5NUN6w+TORJeCxNVfDmFqjo6PdGF/nHqefPeSH/cCKE9OfCGHuYbVGh2Crxboh/OjYLIo0bQQnGmEn8uT4VTXOs9GFP50HSdw503qMIpJ9CRR/fg==
ARC-Authentication-Results: i=1; mx.microsoft.com 1; spf=pass
 smtp.mailfrom=amd.com; dmarc=pass action=none header.from=amd.com; dkim=pass
 header.d=amd.com; arc=none
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
 d=amdcloud.onmicrosoft.com; s=selector2-amdcloud-onmicrosoft-com;
 h=From:Date:Subject:Message-ID:Content-Type:MIME-Version:X-MS-Exchange-SenderADCheck;
 bh=MJ+BpJng5csN/LA+I/qcPQkYM/tISIORmRx+LOYMnks=;
 b=0exTJDH5sq9zBSGZtUC2wPy38PRzvE4az0GJIFUHwASHOsbf89iGW4bYdBmef54a2zCF6MOLfqwlyFW9uPHGYetS1psxUp6mTeIjjEQbr19rHTYePfAoCmy/ZaBBFGzZYYWFAmiG+jhVCfpHSJzOJ8HWlC2k1sOAmmRIjoSHPU0=
Authentication-Results: nongnu.org; dkim=none (message not signed)
 header.d=none;nongnu.org; dmarc=none action=none header.from=amd.com;
Received: from CY4PR12MB1352.namprd12.prod.outlook.com (2603:10b6:903:3a::13)
 by CY4PR12MB1607.namprd12.prod.outlook.com (2603:10b6:910:b::14) with
 Microsoft SMTP Server (version=TLS1_2,
 cipher=TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384) id 15.20.3370.16; Tue, 15 Sep
 2020 21:30:08 +0000
Received: from CY4PR12MB1352.namprd12.prod.outlook.com
 ([fe80::989b:b1b2:464c:443]) by CY4PR12MB1352.namprd12.prod.outlook.com
 ([fe80::989b:b1b2:464c:443%10]) with mapi id 15.20.3370.019; Tue, 15 Sep 2020
 21:30:08 +0000
From: Tom Lendacky <thomas.lendacky@amd.com>
To: qemu-devel@nongnu.org, kvm@vger.kernel.org
Cc: Marcel Apfelbaum <marcel.apfelbaum@gmail.com>,
        Paolo Bonzini <pbonzini@redhat.com>,
        "Dr. David Alan Gilbert" <dgilbert@redhat.com>,
        Eduardo Habkost <ehabkost@redhat.com>,
        Richard Henderson <rth@twiddle.net>,
        Connor Kuehl <ckuehl@redhat.com>,
        Brijesh Singh <brijesh.singh@amd.com>,
        Jiri Slaby <jslaby@suse.cz>,
        Marcelo Tosatti <mtosatti@redhat.com>,
        "Michael S. Tsirkin" <mst@redhat.com>
Subject: [PATCH v3 1/5] sev/i386: Add initial support for SEV-ES
Date: Tue, 15 Sep 2020 16:29:40 -0500
Message-Id: 
 <e2456cc461f329f52aa6eb3fcd0d0ce9451b8fa7.1600205384.git.thomas.lendacky@amd.com>
X-Mailer: git-send-email 2.28.0
In-Reply-To: <cover.1600205384.git.thomas.lendacky@amd.com>
References: <cover.1600205384.git.thomas.lendacky@amd.com>
X-ClientProxiedBy: DM5PR21CA0022.namprd21.prod.outlook.com
 (2603:10b6:3:ac::32) To CY4PR12MB1352.namprd12.prod.outlook.com
 (2603:10b6:903:3a::13)
MIME-Version: 1.0
X-MS-Exchange-MessageSentRepresentingType: 1
Received: from tlendack-t1.amd.com (165.204.77.1) by
 DM5PR21CA0022.namprd21.prod.outlook.com (2603:10b6:3:ac::32) with Microsoft
 SMTP Server (version=TLS1_2,
 cipher=TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384) id 15.20.3412.0 via Frontend
 Transport; Tue, 15 Sep 2020 21:30:07 +0000
X-Mailer: git-send-email 2.28.0
X-Originating-IP: [165.204.77.1]
X-MS-PublicTrafficType: Email
X-MS-Office365-Filtering-HT: Tenant
X-MS-Office365-Filtering-Correlation-Id: 7be9caef-abf7-4671-d990-08d859be84b6
X-MS-TrafficTypeDiagnostic: CY4PR12MB1607:
X-MS-Exchange-Transport-Forked: True
X-Microsoft-Antispam-PRVS: 
 <CY4PR12MB16078003DCFFEDF511B73192EC200@CY4PR12MB1607.namprd12.prod.outlook.com>
X-MS-Oob-TLC-OOBClassifiers: OLM:2043;
X-MS-Exchange-SenderADCheck: 1
X-Microsoft-Antispam: BCL:0;
X-Microsoft-Antispam-Message-Info: 
 42NvKcwMhV+hkZ6a2o5icitj4a+1HG/0kwvwHUR5mq3zfejADqEhR/WTrUuvRQubbEbmEdMxGADIn1G/cY+fl7VPtH+np3xsl8zvlBYKOe8Pxk8tiADOA0BCOMZWhRf7KkvI3xc3vCn3OnIMc94lnCuH1/u1jUR2s9wToHMpy7Xmdw3ajG18vT0wKFfEiaKgAWPf1OnpbwpIbKkKDC6Mp0MJkij+G8H9sDVF00rja/OIPwzI5pfJ8HRIgJfv4Bn2M/dJ/+ib/VH+WJ+VLofALNQ8Q6YXcLtX/oWt5xxQ4DZQnWgKdkiwqLimF2ikzP/F/wFSoVfLLOru8X9y8G7KLKYKsXaGfpY5u1wnTiELRrKDXecHRH9KJPWOq3HF9XvX
X-Forefront-Antispam-Report: 
 CIP:255.255.255.255;CTRY:;LANG:en;SCL:1;SRV:;IPV:NLI;SFV:NSPM;H:CY4PR12MB1352.namprd12.prod.outlook.com;PTR:;CAT:NONE;SFS:(4636009)(396003)(366004)(346002)(136003)(39860400002)(376002)(5660300002)(956004)(478600001)(6666004)(86362001)(52116002)(2616005)(2906002)(7696005)(36756003)(66476007)(8936002)(66556008)(66946007)(316002)(54906003)(7416002)(26005)(6486002)(4326008)(8676002)(16526019)(83380400001)(186003)(309714004);DIR:OUT;SFP:1101;
X-MS-Exchange-AntiSpam-MessageData: 
 I051YFqFnSwiRy6dax9lhZVvzOPLBskuC1xd9oE2mwmBBsvAA4dp0MGUiIfTsYzRY8/pc8rR3DgrnfmiGocgjVCsHcQ/M1rOYQ1VLywqnktyNQWihODVBxywhFC7PI+9w38mmo5gVsJH2Sn3/PlkZ+fq+aWGlXRFJDVQ9jKBM1iCBW8y3jXj8hzOQSymcSc1QW9ANXtWIivN3IS9hU3CKTFzaCo5In9/j6+kk1i0+KLgEWiUZ54RITxlmoqwpi2pefMhBVrxYR81Rh9Fc3mndayjE4ZFY9vv6Y/RuzYnqiJqu872gBglBlo6O5L4mbguFp2B30zQxKLtAj9u1FXTnBS9Ao+w0l6mJ7i2eAD3uCm782dHuiiR6Ec76SbuYGJ562u+TBGiIe1cAohXxW5RbLO6lqZLvEeFRnBc2q8uAdDyvwBLEOJvDeQ6h71B1BgB15qk/ZNAdv9NYBnRc7ZJK7iVHghaY+Ld9I6BTOiB8Dc6IURZt2oiXFvJVlNNBKn+DQd/hUBmq97BG9F0fzc7rVX5U4tZ6vL+yvxew5fhb74EQrpn2w6rLY32AqEVCvEj+ZeaDsAvaulSBq8CPJrha709wxpitDtki4EyD9jm1iw9C+fProVruraDH/IMu6rqH1lLNrL5iNhFhrBoZnLkPQ==
X-OriginatorOrg: amd.com
X-MS-Exchange-CrossTenant-Network-Message-Id: 
 7be9caef-abf7-4671-d990-08d859be84b6
X-MS-Exchange-CrossTenant-AuthSource: CY4PR12MB1352.namprd12.prod.outlook.com
X-MS-Exchange-CrossTenant-AuthAs: Internal
X-MS-Exchange-CrossTenant-OriginalArrivalTime: 15 Sep 2020 21:30:08.0318
 (UTC)
X-MS-Exchange-CrossTenant-FromEntityHeader: Hosted
X-MS-Exchange-CrossTenant-Id: 3dd8961f-e488-4e60-8e11-a82d994e183d
X-MS-Exchange-CrossTenant-MailboxType: HOSTED
X-MS-Exchange-CrossTenant-UserPrincipalName: 
 m3FA5ZZ4wdtW26EPYDOsjn4RbhO0jjfzZyMcQaABeu7xWAoBkyqxfreJxS9WXzZ85CW2kfqMHp/h253jxV108Q==
X-MS-Exchange-Transport-CrossTenantHeadersStamped: CY4PR12MB1607
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org

From: Tom Lendacky <thomas.lendacky@amd.com>

Provide initial support for SEV-ES. This includes creating a function to
indicate the guest is an SEV-ES guest (which will return false until all
support is in place), performing the proper SEV initialization and
ensuring that the guest CPU state is measured as part of the launch.

Co-developed-by: Jiri Slaby <jslaby@suse.cz>
Signed-off-by: Jiri Slaby <jslaby@suse.cz>
Signed-off-by: Tom Lendacky <thomas.lendacky@amd.com>
---
 target/i386/cpu.c      |  1 +
 target/i386/sev-stub.c |  5 +++++
 target/i386/sev.c      | 46 ++++++++++++++++++++++++++++++++++++++++--
 target/i386/sev_i386.h |  1 +
 4 files changed, 51 insertions(+), 2 deletions(-)

diff --git a/target/i386/cpu.c b/target/i386/cpu.c
index 588f32e136..bbbe581d35 100644
--- a/target/i386/cpu.c
+++ b/target/i386/cpu.c
@@ -5969,6 +5969,7 @@ void cpu_x86_cpuid(CPUX86State *env, uint32_t index, uint32_t count,
         break;
     case 0x8000001F:
         *eax = sev_enabled() ? 0x2 : 0;
+        *eax |= sev_es_enabled() ? 0x8 : 0;
         *ebx = sev_get_cbit_position();
         *ebx |= sev_get_reduced_phys_bits() << 6;
         *ecx = 0;
diff --git a/target/i386/sev-stub.c b/target/i386/sev-stub.c
index 88e3f39a1e..040ac90563 100644
--- a/target/i386/sev-stub.c
+++ b/target/i386/sev-stub.c
@@ -49,3 +49,8 @@ SevCapability *sev_get_capabilities(Error **errp)
     error_setg(errp, "SEV is not available in this QEMU");
     return NULL;
 }
+
+bool sev_es_enabled(void)
+{
+    return false;
+}
diff --git a/target/i386/sev.c b/target/i386/sev.c
index c3ecf86704..6c9cd0854b 100644
--- a/target/i386/sev.c
+++ b/target/i386/sev.c
@@ -359,6 +359,12 @@ sev_enabled(void)
     return !!sev_guest;
 }
 
+bool
+sev_es_enabled(void)
+{
+    return false;
+}
+
 uint64_t
 sev_get_me_mask(void)
 {
@@ -578,6 +584,22 @@ sev_launch_update_data(SevGuestState *sev, uint8_t *addr, uint64_t len)
     return ret;
 }
 
+static int
+sev_launch_update_vmsa(SevGuestState *sev)
+{
+    int ret, fw_error;
+
+    ret = sev_ioctl(sev->sev_fd, KVM_SEV_LAUNCH_UPDATE_VMSA, NULL, &fw_error);
+    if (ret) {
+        error_report("%s: LAUNCH_UPDATE_VMSA ret=%d fw_error=%d '%s'",
+                __func__, ret, fw_error, fw_error_to_str(fw_error));
+        goto err;
+    }
+
+err:
+    return ret;
+}
+
 static void
 sev_launch_get_measure(Notifier *notifier, void *unused)
 {
@@ -590,6 +612,14 @@ sev_launch_get_measure(Notifier *notifier, void *unused)
         return;
     }
 
+    if (sev_es_enabled()) {
+        /* measure all the VM save areas before getting launch_measure */
+        ret = sev_launch_update_vmsa(sev);
+        if (ret) {
+            exit(1);
+        }
+    }
+
     measurement = g_new0(struct kvm_sev_launch_measure, 1);
 
     /* query the measurement blob length */
@@ -684,7 +714,7 @@ sev_guest_init(const char *id)
 {
     SevGuestState *sev;
     char *devname;
-    int ret, fw_error;
+    int ret, fw_error, cmd;
     uint32_t ebx;
     uint32_t host_cbitpos;
     struct sev_user_data_status status = {};
@@ -745,8 +775,20 @@ sev_guest_init(const char *id)
     sev->api_major = status.api_major;
     sev->api_minor = status.api_minor;
 
+    if (sev_es_enabled()) {
+        if (!(status.flags & SEV_STATUS_FLAGS_CONFIG_ES)) {
+            error_report("%s: guest policy requires SEV-ES, but "
+                         "host SEV-ES support unavailable",
+                         __func__);
+            goto err;
+        }
+        cmd = KVM_SEV_ES_INIT;
+    } else {
+        cmd = KVM_SEV_INIT;
+    }
+
     trace_kvm_sev_init();
-    ret = sev_ioctl(sev->sev_fd, KVM_SEV_INIT, NULL, &fw_error);
+    ret = sev_ioctl(sev->sev_fd, cmd, NULL, &fw_error);
     if (ret) {
         error_report("%s: failed to initialize ret=%d fw_error=%d '%s'",
                      __func__, ret, fw_error, fw_error_to_str(fw_error));
diff --git a/target/i386/sev_i386.h b/target/i386/sev_i386.h
index 4db6960f60..4f9a5e9b21 100644
--- a/target/i386/sev_i386.h
+++ b/target/i386/sev_i386.h
@@ -29,6 +29,7 @@
 #define SEV_POLICY_SEV          0x20
 
 extern bool sev_enabled(void);
+extern bool sev_es_enabled(void);
 extern uint64_t sev_get_me_mask(void);
 extern SevInfo *sev_get_info(void);
 extern uint32_t sev_get_cbit_position(void);

From patchwork Tue Sep 15 21:29:41 2020
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Tom Lendacky <thomas.lendacky@amd.com>
X-Patchwork-Id: 11777875
Return-Path: <SRS0=V1hm=CY=vger.kernel.org=kvm-owner@kernel.org>
Received: from mail.kernel.org (pdx-korg-mail-1.web.codeaurora.org
 [172.30.200.123])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id ACBF0139F
	for <patchwork-kvm@patchwork.kernel.org>;
 Tue, 15 Sep 2020 21:42:19 +0000 (UTC)
Received: from vger.kernel.org (vger.kernel.org [23.128.96.18])
	by mail.kernel.org (Postfix) with ESMTP id 8BE7820872
	for <patchwork-kvm@patchwork.kernel.org>;
 Tue, 15 Sep 2020 21:42:19 +0000 (UTC)
Authentication-Results: mail.kernel.org;
	dkim=pass (1024-bit key) header.d=amdcloud.onmicrosoft.com
 header.i=@amdcloud.onmicrosoft.com header.b="Uk6gszkL"
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1728030AbgIOVbv (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Tue, 15 Sep 2020 17:31:51 -0400
Received: from mail-bn8nam08on2083.outbound.protection.outlook.com
 ([40.107.100.83]:29249
        "EHLO NAM04-BN8-obe.outbound.protection.outlook.com"
        rhost-flags-OK-OK-OK-FAIL) by vger.kernel.org with ESMTP
        id S1728080AbgIOVav (ORCPT <rfc822;kvm@vger.kernel.org>);
        Tue, 15 Sep 2020 17:30:51 -0400
ARC-Seal: i=1; a=rsa-sha256; s=arcselector9901; d=microsoft.com; cv=none;
 b=k4wz7gsGCwHH3N6SYOFksxI722vG6OONX8HDR3lWkp14pi6B8MZcWuc9oUWRDH3hoZ9efPawV4F1wIX1KkMG301DIr6lhmPODQGtByLyZvQtZKRmRioex2vK0vJFpIsAUTWb9lGx3ozRIvaEFpTyb2sfcXFnQB6cDzKb/taxlHnnrtXyvQyeqjsqVZNwJ2LkkCcKepgBncSxAiTZ/TfMhRnPjXbnXP3K7vfeODLGuj/k2KUxxgH8oDTYEd665WQz8yqcXC0NAYcoaRRlI2hj1P5FUEmAnkWyg+Uo1AZtERvKkdxRBEWh69TA6Kp75YhO39bHtq+H+BDxGD1Splt8iA==
ARC-Message-Signature: i=1; a=rsa-sha256; c=relaxed/relaxed; d=microsoft.com;
 s=arcselector9901;
 h=From:Date:Subject:Message-ID:Content-Type:MIME-Version:X-MS-Exchange-SenderADCheck;
 bh=pwUL4IAYHpyXvJJHQEV+j4J5TbqEfkYMfCqoG7chKXE=;
 b=PTx4WhBNnGM89Z+HJq9pQcMgrXy4mYPnLp9GGRDTVJYQyDI9ec7AWOWG62wGpkDnd3lIgqRVMohXL4yKOtiGfopQFjvV58lkzQSgMjnukPTz464VKSdJbNsFn+4vI2+WYHSEd+0k4UaX3g7+VJkxvCyoc0zhHnCKGIovHRkOpdpP7BMuvaxDxd1Atezs9c0sVgpDpJp6gsY3M1yS6cnxRy6sz6EYTnWq23uE+YytV9xNFoQ6ujXZllnD9Vcm4yS2iSZEAsYQWV0XEwhMGVrQuCyYZWNEJGLlBi7ZEOlT0RBVTigSb/hKtLRw69lzVb8WpIq5C4FyWSJJQ9HkooqyWQ==
ARC-Authentication-Results: i=1; mx.microsoft.com 1; spf=pass
 smtp.mailfrom=amd.com; dmarc=pass action=none header.from=amd.com; dkim=pass
 header.d=amd.com; arc=none
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
 d=amdcloud.onmicrosoft.com; s=selector2-amdcloud-onmicrosoft-com;
 h=From:Date:Subject:Message-ID:Content-Type:MIME-Version:X-MS-Exchange-SenderADCheck;
 bh=pwUL4IAYHpyXvJJHQEV+j4J5TbqEfkYMfCqoG7chKXE=;
 b=Uk6gszkL/eDiX8tXdtJD1rlwe2+9yj7h09zBidK0zPQbu2UOvZ4CjOd42TfO1s7llva1IQtaT0Yy62Qdm9MIzZn9cWuf6AaDPMZ8ioEra5AL+Wrsc7CuuKAGfJBo13oBQbP2Q50sXWT2a+9ncrrSSesjMr4oDOnZM39RtSvE/ow=
Authentication-Results: nongnu.org; dkim=none (message not signed)
 header.d=none;nongnu.org; dmarc=none action=none header.from=amd.com;
Received: from CY4PR12MB1352.namprd12.prod.outlook.com (2603:10b6:903:3a::13)
 by CY4PR12MB1607.namprd12.prod.outlook.com (2603:10b6:910:b::14) with
 Microsoft SMTP Server (version=TLS1_2,
 cipher=TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384) id 15.20.3370.16; Tue, 15 Sep
 2020 21:30:16 +0000
Received: from CY4PR12MB1352.namprd12.prod.outlook.com
 ([fe80::989b:b1b2:464c:443]) by CY4PR12MB1352.namprd12.prod.outlook.com
 ([fe80::989b:b1b2:464c:443%10]) with mapi id 15.20.3370.019; Tue, 15 Sep 2020
 21:30:16 +0000
From: Tom Lendacky <thomas.lendacky@amd.com>
To: qemu-devel@nongnu.org, kvm@vger.kernel.org
Cc: Marcel Apfelbaum <marcel.apfelbaum@gmail.com>,
        Paolo Bonzini <pbonzini@redhat.com>,
        "Dr. David Alan Gilbert" <dgilbert@redhat.com>,
        Eduardo Habkost <ehabkost@redhat.com>,
        Richard Henderson <rth@twiddle.net>,
        Connor Kuehl <ckuehl@redhat.com>,
        Brijesh Singh <brijesh.singh@amd.com>,
        Jiri Slaby <jslaby@suse.cz>,
        Marcelo Tosatti <mtosatti@redhat.com>,
        "Michael S. Tsirkin" <mst@redhat.com>
Subject: [PATCH v3 2/5] sev/i386: Require in-kernel irqchip support for SEV-ES
 guests
Date: Tue, 15 Sep 2020 16:29:41 -0500
Message-Id: 
 <59ca62e1527a854fc8c411fb01bf0f55ee13aeb7.1600205384.git.thomas.lendacky@amd.com>
X-Mailer: git-send-email 2.28.0
In-Reply-To: <cover.1600205384.git.thomas.lendacky@amd.com>
References: <cover.1600205384.git.thomas.lendacky@amd.com>
X-ClientProxiedBy: DM5PR21CA0053.namprd21.prod.outlook.com
 (2603:10b6:3:129::15) To CY4PR12MB1352.namprd12.prod.outlook.com
 (2603:10b6:903:3a::13)
MIME-Version: 1.0
X-MS-Exchange-MessageSentRepresentingType: 1
Received: from tlendack-t1.amd.com (165.204.77.1) by
 DM5PR21CA0053.namprd21.prod.outlook.com (2603:10b6:3:129::15) with Microsoft
 SMTP Server (version=TLS1_2,
 cipher=TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384) id 15.20.3412.1 via Frontend
 Transport; Tue, 15 Sep 2020 21:30:15 +0000
X-Mailer: git-send-email 2.28.0
X-Originating-IP: [165.204.77.1]
X-MS-PublicTrafficType: Email
X-MS-Office365-Filtering-HT: Tenant
X-MS-Office365-Filtering-Correlation-Id: 8434a1a4-dec5-43ea-d2bf-08d859be89be
X-MS-TrafficTypeDiagnostic: CY4PR12MB1607:
X-MS-Exchange-Transport-Forked: True
X-Microsoft-Antispam-PRVS: 
 <CY4PR12MB16071FCE9B025799E42F3805EC200@CY4PR12MB1607.namprd12.prod.outlook.com>
X-MS-Oob-TLC-OOBClassifiers: OLM:989;
X-MS-Exchange-SenderADCheck: 1
X-Microsoft-Antispam: BCL:0;
X-Microsoft-Antispam-Message-Info: 
 /tgiEkMy1imv5MAlhFkAr+Dsdtwys5kLUDHz8xALlQK4kP3Uq/0zhZfGVWntmiKw2nu3U4vuf6jN8ZwZsifkhAe+ytuFUNPKd8xSAuITC6G9iVg9GNpeUVq/AsE20UX9jGMEOxWC6pa1qR5/YALgQL90VJLdtm/WeT+2S6tRGPGRqf8rHsQ5DiKahtkkCytHJWWTB6rYH60k5l7Q8FFVwXF/P22iqWuNXhcCt42XX/7DB8EIsYcD7Rmhow/s4F5MZYuFyAfot6jcpqG4l8wLdT+dpKp0rdQYkSXeUPO+TQPjeY6UarsvXnDEsvs78txT
X-Forefront-Antispam-Report: 
 CIP:255.255.255.255;CTRY:;LANG:en;SCL:1;SRV:;IPV:NLI;SFV:NSPM;H:CY4PR12MB1352.namprd12.prod.outlook.com;PTR:;CAT:NONE;SFS:(4636009)(396003)(366004)(346002)(136003)(39860400002)(376002)(5660300002)(956004)(478600001)(6666004)(86362001)(52116002)(2616005)(2906002)(7696005)(36756003)(66476007)(8936002)(66556008)(66946007)(4744005)(316002)(54906003)(7416002)(26005)(6486002)(4326008)(8676002)(16526019)(186003);DIR:OUT;SFP:1101;
X-MS-Exchange-AntiSpam-MessageData: 
 FGZPTAE35glA00aZnnUUzL6GK6EP11xfDDzr6f/XFCsj8vDhBm6ghCxIFoaoU2VdvLe1uCBsnT5upU5BNFC1/uZFzE1t1khr/Juw5p0uTIDvr8OUV1C74mOkyjnBUUsfUwuaotKqe0GZHWSRiEMPtRMZBE1jikXepWOBDT1EsgNtTdy+wZQcYPZ4Am3fBgwgJbw4eJ5jfEUSmdXDSeabcyh3wz9ur/zVH902yflj7JXErijo7uU93g2n9ycoU0MImIW7NhlgkWAinIxtUaw3Bx9FRtDUqtIB1YF49OEEVqQVpse0dq9NjQ4o2hB/7yNkajSqrDgCPr6JHz03ge8cKOZYQV2Zscj+0SstldN2xAG9xU/1GsnzDx99LvpePHdqTijgQKcZ6EqxlsMMyi30IjaZMKYbN55ay24Kvde+GbtNVh7Ix+cCPqPsZEevqzIz6i3VAdpCsZqYRiXlGKyTVuUhRaQWY/N9RNGF/z/ylrplPIaD5/+m3L6McsX6xa0THFkolKvsIJqTxMoed79xuTW8WDqToVjktd76NrDF+PxJn4hgr1rZt2wsRB+pvZlfV3STe1w2P0f1de/w5PErJ+vpfptN5bXOcc77r8wct5wjUXfK1MTlx+qFusbtuO6mDDVNw72dgSgVyr2myfWbxA==
X-OriginatorOrg: amd.com
X-MS-Exchange-CrossTenant-Network-Message-Id: 
 8434a1a4-dec5-43ea-d2bf-08d859be89be
X-MS-Exchange-CrossTenant-AuthSource: CY4PR12MB1352.namprd12.prod.outlook.com
X-MS-Exchange-CrossTenant-AuthAs: Internal
X-MS-Exchange-CrossTenant-OriginalArrivalTime: 15 Sep 2020 21:30:16.4816
 (UTC)
X-MS-Exchange-CrossTenant-FromEntityHeader: Hosted
X-MS-Exchange-CrossTenant-Id: 3dd8961f-e488-4e60-8e11-a82d994e183d
X-MS-Exchange-CrossTenant-MailboxType: HOSTED
X-MS-Exchange-CrossTenant-UserPrincipalName: 
 dQsfTw+AAXbg2cFx+5hJ/M8WbNVJyKfTB8MOUXVCb14SfhNE55IdzZ4h5z9IZLnx2wPQljr8DneArV5WV+o19A==
X-MS-Exchange-Transport-CrossTenantHeadersStamped: CY4PR12MB1607
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org

From: Tom Lendacky <thomas.lendacky@amd.com>

In prep for AP booting, require the use of in-kernel irqchip support. This
lessens the Qemu support burden required to boot APs.

Signed-off-by: Tom Lendacky <thomas.lendacky@amd.com>
---
 target/i386/sev.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/target/i386/sev.c b/target/i386/sev.c
index 6c9cd0854b..5055b1fe00 100644
--- a/target/i386/sev.c
+++ b/target/i386/sev.c
@@ -776,6 +776,12 @@ sev_guest_init(const char *id)
     sev->api_minor = status.api_minor;
 
     if (sev_es_enabled()) {
+        if (!kvm_kernel_irqchip_allowed()) {
+            error_report("%s: SEV-ES guests require in-kernel irqchip support",
+                         __func__);
+            goto err;
+        }
+
         if (!(status.flags & SEV_STATUS_FLAGS_CONFIG_ES)) {
             error_report("%s: guest policy requires SEV-ES, but "
                          "host SEV-ES support unavailable",

From patchwork Tue Sep 15 21:29:42 2020
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Tom Lendacky <thomas.lendacky@amd.com>
X-Patchwork-Id: 11777879
Return-Path: <SRS0=V1hm=CY=vger.kernel.org=kvm-owner@kernel.org>
Received: from mail.kernel.org (pdx-korg-mail-1.web.codeaurora.org
 [172.30.200.123])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id C5A39618
	for <patchwork-kvm@patchwork.kernel.org>;
 Tue, 15 Sep 2020 21:43:09 +0000 (UTC)
Received: from vger.kernel.org (vger.kernel.org [23.128.96.18])
	by mail.kernel.org (Postfix) with ESMTP id A0C7520809
	for <patchwork-kvm@patchwork.kernel.org>;
 Tue, 15 Sep 2020 21:43:09 +0000 (UTC)
Authentication-Results: mail.kernel.org;
	dkim=pass (1024-bit key) header.d=amdcloud.onmicrosoft.com
 header.i=@amdcloud.onmicrosoft.com header.b="CsBYkv7J"
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1728093AbgIOVm5 (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Tue, 15 Sep 2020 17:42:57 -0400
Received: from mail-dm6nam10on2059.outbound.protection.outlook.com
 ([40.107.93.59]:21344
        "EHLO NAM10-DM6-obe.outbound.protection.outlook.com"
        rhost-flags-OK-OK-OK-FAIL) by vger.kernel.org with ESMTP
        id S1728042AbgIOVbf (ORCPT <rfc822;kvm@vger.kernel.org>);
        Tue, 15 Sep 2020 17:31:35 -0400
ARC-Seal: i=1; a=rsa-sha256; s=arcselector9901; d=microsoft.com; cv=none;
 b=jUfJrTYUwym3pep1eiJ8JZ/hfVN2yRy2LJwBGcfTC+naq/ZAm3L8MP5ezN7EO/GpkTN9xMrz0uOyoJsyseEKraXXv1e9BI2Ts89nWFnFhuI7YD1VJGPNBqsXkSHbN5TMdziOSBqu5/GicIq7A2kZ5bupIo9fjKdpDxz/zK0Qp3xsc7NZli9YpYeUcSTfPy0FsS5jkxlC8plRLlSRz/HzrCMPu8YscqRYDNvAWF3JOJpIu955Z5h+CClKPa2Hr/5J1E+V32C3FOT77LhaeBxBa77sr1D2XJ7G8OnU90HQ9Z+ZuTw4uQy31T53G7lcM/DA/oQb+rNzM2zZkMrLDBA2WA==
ARC-Message-Signature: i=1; a=rsa-sha256; c=relaxed/relaxed; d=microsoft.com;
 s=arcselector9901;
 h=From:Date:Subject:Message-ID:Content-Type:MIME-Version:X-MS-Exchange-SenderADCheck;
 bh=hFI1iSZIUByoFPbo3zOQC+VVjUYP8EgCKTSOqArNAuw=;
 b=Wldf0+x6VTWbsaNh14ILLrUlaeq7gYUw3iyYuJmkz2u49RYHWAVChYyyFoWgfehW6MjCdMI0G+EcVY6dzmwLNWM9pnGq2+d+Vjj624Grlj5C7DeXGrhT4YjlYUeghwQxCV7L9vt0Lpp8LUycBhE5j8SvwuTiFmrtO0Z7ujIhnoKmkPG+r4tqOzOZY3xpx0TffI4/a2XVr4Q+R1bGyWkatCQyjjasK23Ka47YBIdP/Ir3UwWeVJItRyCcXonPWjaezSQzuU1LZiKgLJDKO91NU57p10jdKKvhNYHHy4ncCPmeh4jOVktG96Can78e3CfDLxkjos8WIjgYDqUMC1u1Kw==
ARC-Authentication-Results: i=1; mx.microsoft.com 1; spf=pass
 smtp.mailfrom=amd.com; dmarc=pass action=none header.from=amd.com; dkim=pass
 header.d=amd.com; arc=none
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
 d=amdcloud.onmicrosoft.com; s=selector2-amdcloud-onmicrosoft-com;
 h=From:Date:Subject:Message-ID:Content-Type:MIME-Version:X-MS-Exchange-SenderADCheck;
 bh=hFI1iSZIUByoFPbo3zOQC+VVjUYP8EgCKTSOqArNAuw=;
 b=CsBYkv7JgH5zbsyhg+OtUIcBMVaAbB048mSv+lwt0nT4dKseQwc6KreDeDH0I2AL52DmwuOe3wvTWIZPZBD/aTIJ9F3Qti/wTSDv1NvtlX1VRghjDQwJwpinmtckvYTXB/n4w7PJLMPCPTbu0LNGntFua1uBPmlJGAZ52bR7goU=
Authentication-Results: nongnu.org; dkim=none (message not signed)
 header.d=none;nongnu.org; dmarc=none action=none header.from=amd.com;
Received: from CY4PR12MB1352.namprd12.prod.outlook.com (2603:10b6:903:3a::13)
 by CY4PR12MB1607.namprd12.prod.outlook.com (2603:10b6:910:b::14) with
 Microsoft SMTP Server (version=TLS1_2,
 cipher=TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384) id 15.20.3370.16; Tue, 15 Sep
 2020 21:30:26 +0000
Received: from CY4PR12MB1352.namprd12.prod.outlook.com
 ([fe80::989b:b1b2:464c:443]) by CY4PR12MB1352.namprd12.prod.outlook.com
 ([fe80::989b:b1b2:464c:443%10]) with mapi id 15.20.3370.019; Tue, 15 Sep 2020
 21:30:25 +0000
From: Tom Lendacky <thomas.lendacky@amd.com>
To: qemu-devel@nongnu.org, kvm@vger.kernel.org
Cc: Marcel Apfelbaum <marcel.apfelbaum@gmail.com>,
        Paolo Bonzini <pbonzini@redhat.com>,
        "Dr. David Alan Gilbert" <dgilbert@redhat.com>,
        Eduardo Habkost <ehabkost@redhat.com>,
        Richard Henderson <rth@twiddle.net>,
        Connor Kuehl <ckuehl@redhat.com>,
        Brijesh Singh <brijesh.singh@amd.com>,
        Jiri Slaby <jslaby@suse.cz>,
        Marcelo Tosatti <mtosatti@redhat.com>,
        "Michael S. Tsirkin" <mst@redhat.com>
Subject: [PATCH v3 3/5] sev/i386: Allow AP booting under SEV-ES
Date: Tue, 15 Sep 2020 16:29:42 -0500
Message-Id: 
 <9d964b7575471f45c522eea9ea3a7d84ed4d7d2b.1600205384.git.thomas.lendacky@amd.com>
X-Mailer: git-send-email 2.28.0
In-Reply-To: <cover.1600205384.git.thomas.lendacky@amd.com>
References: <cover.1600205384.git.thomas.lendacky@amd.com>
X-ClientProxiedBy: DM3PR11CA0019.namprd11.prod.outlook.com
 (2603:10b6:0:54::29) To CY4PR12MB1352.namprd12.prod.outlook.com
 (2603:10b6:903:3a::13)
MIME-Version: 1.0
X-MS-Exchange-MessageSentRepresentingType: 1
Received: from tlendack-t1.amd.com (165.204.77.1) by
 DM3PR11CA0019.namprd11.prod.outlook.com (2603:10b6:0:54::29) with Microsoft
 SMTP Server (version=TLS1_2,
 cipher=TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384) id 15.20.3391.11 via Frontend
 Transport; Tue, 15 Sep 2020 21:30:24 +0000
X-Mailer: git-send-email 2.28.0
X-Originating-IP: [165.204.77.1]
X-MS-PublicTrafficType: Email
X-MS-Office365-Filtering-HT: Tenant
X-MS-Office365-Filtering-Correlation-Id: 889c9441-840d-4d20-bf34-08d859be8f1b
X-MS-TrafficTypeDiagnostic: CY4PR12MB1607:
X-MS-Exchange-Transport-Forked: True
X-Microsoft-Antispam-PRVS: 
 <CY4PR12MB160704B9FF19C4EF0DC3C13BEC200@CY4PR12MB1607.namprd12.prod.outlook.com>
X-MS-Oob-TLC-OOBClassifiers: OLM:568;
X-MS-Exchange-SenderADCheck: 1
X-Microsoft-Antispam: BCL:0;
X-Microsoft-Antispam-Message-Info: 
 DCdx+RKlpXbUDYOOwCYqzQVbxmbxngkmdJdlld2hyegym5q+jf+UEw0qnwqQ2u0OrGdIsqkWAvnadrObvx7v2vGJ8Y2SK+D6SzqBPkkjmnWNCLYScIVUa5iEcciYZ2LuK7QcHkhKNkZUYgidLUBkVHDRChL7j8sNWJEozxnz00a8I/KPhs/4BR9qDxbNRmm4b6DpVbJ4y7HWmZWMPmYE8mlHcNY9lkLbDm2Y4GUNZyCIEx2YBwbp6aCNi156FdwWe7bOvnDByFdN5ACqfQHpdTV7PPoyeu5OWfUusrTCR8Ne0keSTdklZ0skzsFXF6smJql7KkKBbvYKoDuCE0D49g==
X-Forefront-Antispam-Report: 
 CIP:255.255.255.255;CTRY:;LANG:en;SCL:1;SRV:;IPV:NLI;SFV:NSPM;H:CY4PR12MB1352.namprd12.prod.outlook.com;PTR:;CAT:NONE;SFS:(4636009)(396003)(366004)(346002)(136003)(39860400002)(376002)(5660300002)(956004)(478600001)(6666004)(86362001)(52116002)(2616005)(2906002)(7696005)(36756003)(66476007)(8936002)(66556008)(66946007)(316002)(54906003)(7416002)(26005)(6486002)(4326008)(8676002)(16526019)(83380400001)(186003);DIR:OUT;SFP:1101;
X-MS-Exchange-AntiSpam-MessageData: 
 jbOl3VtfwWeptjhbDam0tO1idYtO54FSaLIYhwTLzU+MOsviyp0U9K2qDW21fd9IrkGcN+NHCYW4TgzpJiH7pakZm3gmFFK67UcVmnVSucpoEpDIDPuaDTzw2XFBjkH0xqxiQYnC0sOdCglbN3ZqPe7S82JpeD9aPWU+2fbSms+KHx0r6pvMmKT8o7CoEUpnEv1gD0+PxgCCmRCh7ccwVFelRxfsFvQuH/jGWNy5TG6yKFI7m6VMvK5Cw38Yu8JiMEUhluKMtlvozDTtsMHb1w3uprBTLTd7/xp9rdojgASlnMROx+5z4KXlDsBR8iJSrZfvwOd3z8CITELNqf1mJJu3fDfHpb+Lwgr93xLWLWENEmE8+Qlc7hazEWnc+caufJenUPAHkfo6Mf7mlemm35WaDkKM48rnPdDJeGtRelICkdQ3ETspLX4tkPeWRDOtNKcAjHAvJLqcKpRvJWBp883Rnum5wm5n7Hcz2gClVPjdb11En2zkzx0n865f+3mBAFTmEQM7BSuOhiBkE+rDWjmKoNDNopSFXEulhYIgGgNUGQv260VMsss2v6YCY6QjnnCUPwM11Xe0ke1Lz/WN1tru2f/mhZXZ7sHwGeMz2XU10dBRICudi2U+uiFoi0QqRowRfT3hO5W3ytrj5ZjNSw==
X-OriginatorOrg: amd.com
X-MS-Exchange-CrossTenant-Network-Message-Id: 
 889c9441-840d-4d20-bf34-08d859be8f1b
X-MS-Exchange-CrossTenant-AuthSource: CY4PR12MB1352.namprd12.prod.outlook.com
X-MS-Exchange-CrossTenant-AuthAs: Internal
X-MS-Exchange-CrossTenant-OriginalArrivalTime: 15 Sep 2020 21:30:25.5262
 (UTC)
X-MS-Exchange-CrossTenant-FromEntityHeader: Hosted
X-MS-Exchange-CrossTenant-Id: 3dd8961f-e488-4e60-8e11-a82d994e183d
X-MS-Exchange-CrossTenant-MailboxType: HOSTED
X-MS-Exchange-CrossTenant-UserPrincipalName: 
 0zDfGC+nKp8R6fC5MdqW+lG5cP2FUEiLzTWVAuEmE+93p6U+A8M2ZTJDKrXmG33HgbTT1XtKoqU+I19+XAA9ZQ==
X-MS-Exchange-Transport-CrossTenantHeadersStamped: CY4PR12MB1607
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org

From: Tom Lendacky <thomas.lendacky@amd.com>

When SEV-ES is enabled, it is not possible modify the guests register
state after it has been initially created, encrypted and measured.

Normally, an INIT-SIPI-SIPI request is used to boot the AP. However, the
hypervisor cannot emulate this because it cannot update the AP register
state. For the very first boot by an AP, the reset vector CS segment
value and the EIP value must be programmed before the register has been
encrypted and measured.

Signed-off-by: Tom Lendacky <thomas.lendacky@amd.com>
---
 accel/kvm/kvm-all.c    | 64 ++++++++++++++++++++++++++++++++++++++++++
 accel/stubs/kvm-stub.c |  5 ++++
 hw/i386/pc_sysfw.c     | 10 ++++++-
 include/sysemu/kvm.h   | 16 +++++++++++
 include/sysemu/sev.h   |  3 ++
 target/i386/kvm.c      |  2 ++
 target/i386/sev.c      | 51 +++++++++++++++++++++++++++++++++
 7 files changed, 150 insertions(+), 1 deletion(-)

diff --git a/accel/kvm/kvm-all.c b/accel/kvm/kvm-all.c
index 63ef6af9a1..20725b0368 100644
--- a/accel/kvm/kvm-all.c
+++ b/accel/kvm/kvm-all.c
@@ -39,6 +39,7 @@
 #include "qemu/main-loop.h"
 #include "trace.h"
 #include "hw/irq.h"
+#include "sysemu/kvm.h"
 #include "sysemu/sev.h"
 #include "qapi/visitor.h"
 #include "qapi/qapi-types-common.h"
@@ -120,6 +121,12 @@ struct KVMState
     /* memory encryption */
     void *memcrypt_handle;
     int (*memcrypt_encrypt_data)(void *handle, uint8_t *ptr, uint64_t len);
+    int (*memcrypt_save_reset_vector)(void *handle, void *flash_ptr,
+                                      uint64_t flash_size, uint32_t *addr);
+
+    unsigned int reset_cs;
+    unsigned int reset_ip;
+    bool reset_valid;
 
     /* For "info mtree -f" to tell if an MR is registered in KVM */
     int nr_as;
@@ -239,6 +246,62 @@ int kvm_memcrypt_encrypt_data(uint8_t *ptr, uint64_t len)
     return 1;
 }
 
+void kvm_memcrypt_set_reset_vector(CPUState *cpu)
+{
+    X86CPU *x86;
+    CPUX86State *env;
+
+    /* Only update if we have valid reset information */
+    if (!kvm_state->reset_valid) {
+        return;
+    }
+
+    /* Do not update the BSP reset state */
+    if (cpu->cpu_index == 0) {
+        return;
+    }
+
+    x86 = X86_CPU(cpu);
+    env = &x86->env;
+
+    cpu_x86_load_seg_cache(env, R_CS, 0xf000, kvm_state->reset_cs, 0xffff,
+                           DESC_P_MASK | DESC_S_MASK | DESC_CS_MASK |
+                           DESC_R_MASK | DESC_A_MASK);
+
+    env->eip = kvm_state->reset_ip;
+}
+
+int kvm_memcrypt_save_reset_vector(void *flash_ptr, uint64_t flash_size)
+{
+    CPUState *cpu;
+    uint32_t addr;
+    int ret;
+
+    if (kvm_memcrypt_enabled() &&
+        kvm_state->memcrypt_save_reset_vector) {
+
+        addr = 0;
+        ret = kvm_state->memcrypt_save_reset_vector(kvm_state->memcrypt_handle,
+                                                    flash_ptr, flash_size,
+                                                    &addr);
+        if (ret) {
+            return ret;
+        }
+
+        if (addr) {
+            kvm_state->reset_cs = addr & 0xffff0000;
+            kvm_state->reset_ip = addr & 0x0000ffff;
+            kvm_state->reset_valid = true;
+
+            CPU_FOREACH(cpu) {
+                kvm_memcrypt_set_reset_vector(cpu);
+            }
+        }
+    }
+
+    return 0;
+}
+
 /* Called with KVMMemoryListener.slots_lock held */
 static KVMSlot *kvm_get_free_slot(KVMMemoryListener *kml)
 {
@@ -2193,6 +2256,7 @@ static int kvm_init(MachineState *ms)
         }
 
         kvm_state->memcrypt_encrypt_data = sev_encrypt_data;
+        kvm_state->memcrypt_save_reset_vector = sev_es_save_reset_vector;
     }
 
     ret = kvm_arch_init(ms, s);
diff --git a/accel/stubs/kvm-stub.c b/accel/stubs/kvm-stub.c
index 82f118d2df..3aece9b513 100644
--- a/accel/stubs/kvm-stub.c
+++ b/accel/stubs/kvm-stub.c
@@ -114,6 +114,11 @@ int kvm_memcrypt_encrypt_data(uint8_t *ptr, uint64_t len)
   return 1;
 }
 
+int kvm_memcrypt_save_reset_vector(void *flash_ptr, uint64_t flash_size)
+{
+    return -ENOSYS;
+}
+
 #ifndef CONFIG_USER_ONLY
 int kvm_irqchip_add_msi_route(KVMState *s, int vector, PCIDevice *dev)
 {
diff --git a/hw/i386/pc_sysfw.c b/hw/i386/pc_sysfw.c
index b6c0822fe3..321ff94261 100644
--- a/hw/i386/pc_sysfw.c
+++ b/hw/i386/pc_sysfw.c
@@ -156,7 +156,8 @@ static void pc_system_flash_map(PCMachineState *pcms,
     PFlashCFI01 *system_flash;
     MemoryRegion *flash_mem;
     void *flash_ptr;
-    int ret, flash_size;
+    uint64_t flash_size;
+    int ret;
 
     assert(PC_MACHINE_GET_CLASS(pcms)->pci_enabled);
 
@@ -204,6 +205,13 @@ static void pc_system_flash_map(PCMachineState *pcms,
             if (kvm_memcrypt_enabled()) {
                 flash_ptr = memory_region_get_ram_ptr(flash_mem);
                 flash_size = memory_region_size(flash_mem);
+
+                ret = kvm_memcrypt_save_reset_vector(flash_ptr, flash_size);
+                if (ret) {
+                    error_report("failed to locate and/or save reset vector");
+                    exit(1);
+                }
+
                 ret = kvm_memcrypt_encrypt_data(flash_ptr, flash_size);
                 if (ret) {
                     error_report("failed to encrypt pflash rom");
diff --git a/include/sysemu/kvm.h b/include/sysemu/kvm.h
index b4174d941c..f74cfa85ab 100644
--- a/include/sysemu/kvm.h
+++ b/include/sysemu/kvm.h
@@ -247,6 +247,22 @@ bool kvm_memcrypt_enabled(void);
  */
 int kvm_memcrypt_encrypt_data(uint8_t *ptr, uint64_t len);
 
+/**
+ * kvm_memcrypt_set_reset_vector - sets the CS/IP value for the AP if SEV-ES
+ *                                 is active.
+ */
+void kvm_memcrypt_set_reset_vector(CPUState *cpu);
+
+/**
+ * kvm_memcrypt_save_reset_vector - locates and saves the reset vector to be
+ *                                  used as the initial CS/IP value for APs
+ *                                  if SEV-ES is active.
+ *
+ * Return: 1 SEV-ES is active and failed to locate a valid reset vector
+ *         0 SEV-ES is not active or successfully located and saved the
+ *           reset vector address
+ */
+int kvm_memcrypt_save_reset_vector(void *flash_prt, uint64_t flash_size);
 
 #ifdef NEED_CPU_H
 #include "cpu.h"
diff --git a/include/sysemu/sev.h b/include/sysemu/sev.h
index 98c1ec8d38..5198e5a621 100644
--- a/include/sysemu/sev.h
+++ b/include/sysemu/sev.h
@@ -18,4 +18,7 @@
 
 void *sev_guest_init(const char *id);
 int sev_encrypt_data(void *handle, uint8_t *ptr, uint64_t len);
+int sev_es_save_reset_vector(void *handle, void *flash_ptr,
+                             uint64_t flash_size, uint32_t *addr);
+
 #endif
diff --git a/target/i386/kvm.c b/target/i386/kvm.c
index 6f18d940a5..10eaba8943 100644
--- a/target/i386/kvm.c
+++ b/target/i386/kvm.c
@@ -1912,6 +1912,8 @@ void kvm_arch_reset_vcpu(X86CPU *cpu)
     }
     /* enabled by default */
     env->poll_control_msr = 1;
+
+    kvm_memcrypt_set_reset_vector(CPU(cpu));
 }
 
 void kvm_arch_do_init_vcpu(X86CPU *cpu)
diff --git a/target/i386/sev.c b/target/i386/sev.c
index 5055b1fe00..6ddefc65fa 100644
--- a/target/i386/sev.c
+++ b/target/i386/sev.c
@@ -70,6 +70,19 @@ struct SevGuestState {
 #define DEFAULT_GUEST_POLICY    0x1 /* disable debug */
 #define DEFAULT_SEV_DEVICE      "/dev/sev"
 
+/* SEV Information Block GUID = 00f771de-1a7e-4fcb-890e-68c77e2fb44e */
+#define SEV_INFO_BLOCK_GUID \
+    "\xde\x71\xf7\x00\x7e\x1a\xcb\x4f\x89\x0e\x68\xc7\x7e\x2f\xb4\x4e"
+
+typedef struct __attribute__((__packed__)) SevInfoBlock {
+    /* SEV-ES Reset Vector Address */
+    uint32_t reset_addr;
+
+    /* SEV Information Block size and GUID */
+    uint16_t size;
+    char guid[16];
+} SevInfoBlock;
+
 static SevGuestState *sev_guest;
 static Error *sev_mig_blocker;
 
@@ -833,6 +846,44 @@ sev_encrypt_data(void *handle, uint8_t *ptr, uint64_t len)
     return 0;
 }
 
+int
+sev_es_save_reset_vector(void *handle, void *flash_ptr, uint64_t flash_size,
+                         uint32_t *addr)
+{
+    SevInfoBlock *info;
+
+    assert(handle);
+
+    /*
+     * Initialize the address to zero. An address of zero with a successful
+     * return code indicates that SEV-ES is not active.
+     */
+    *addr = 0;
+    if (!sev_es_enabled()) {
+        return 0;
+    }
+
+    /*
+     * Extract the AP reset vector for SEV-ES guests by locating the SEV GUID.
+     * The SEV GUID is located 32 bytes from the end of the flash. Use this
+     * address to base the SEV information block.
+     */
+    info = flash_ptr + flash_size - 0x20 - sizeof(*info);
+    if (memcmp(info->guid, SEV_INFO_BLOCK_GUID, 16)) {
+        error_report("SEV information block not found in pflash rom");
+        return 1;
+    }
+
+    if (!info->reset_addr) {
+        error_report("SEV-ES reset address is zero");
+        return 1;
+    }
+
+    *addr = info->reset_addr;
+
+    return 0;
+}
+
 static void
 sev_register_types(void)
 {

From patchwork Tue Sep 15 21:29:43 2020
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Tom Lendacky <thomas.lendacky@amd.com>
X-Patchwork-Id: 11777885
Return-Path: <SRS0=V1hm=CY=vger.kernel.org=kvm-owner@kernel.org>
Received: from mail.kernel.org (pdx-korg-mail-1.web.codeaurora.org
 [172.30.200.123])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id 539DE14F6
	for <patchwork-kvm@patchwork.kernel.org>;
 Tue, 15 Sep 2020 21:45:08 +0000 (UTC)
Received: from vger.kernel.org (vger.kernel.org [23.128.96.18])
	by mail.kernel.org (Postfix) with ESMTP id 3376020872
	for <patchwork-kvm@patchwork.kernel.org>;
 Tue, 15 Sep 2020 21:45:08 +0000 (UTC)
Authentication-Results: mail.kernel.org;
	dkim=pass (1024-bit key) header.d=amdcloud.onmicrosoft.com
 header.i=@amdcloud.onmicrosoft.com header.b="lO8L1F44"
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1728080AbgIOVor (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Tue, 15 Sep 2020 17:44:47 -0400
Received: from mail-bn8nam08on2083.outbound.protection.outlook.com
 ([40.107.100.83]:29249
        "EHLO NAM04-BN8-obe.outbound.protection.outlook.com"
        rhost-flags-OK-OK-OK-FAIL) by vger.kernel.org with ESMTP
        id S1728085AbgIOVbC (ORCPT <rfc822;kvm@vger.kernel.org>);
        Tue, 15 Sep 2020 17:31:02 -0400
ARC-Seal: i=1; a=rsa-sha256; s=arcselector9901; d=microsoft.com; cv=none;
 b=Ql1yOt/1sjhmOsDzx40ItUm+M0OydLFK446qUaZDraTcRuusQ0ac/Z207poV8ERodka2AHXbLkTrNKUIuPNNYBmepjzTBDcf7FT/Dn1/TUdu6wobue5IzVHyrhRyj0wXLtJuwsjXW0Us8rrIblbk3QloQzWPAz2CiMK4ZgEuMvIcV46/rwX74iRnyNylRmQCDLfgBq8eWm3oPyVTVATr28P2+9taw80i4yzxtIQuSrkMoPI5kJZSXzBDLGI+l/YZEGln/3Ity7bPxGQIS1MYqVKO7d+nVYM5lB7UmfuATb9dm6pSwjxUwi2Ion5+IWgqEMhhtxQS4qWwnz0ek5iUtw==
ARC-Message-Signature: i=1; a=rsa-sha256; c=relaxed/relaxed; d=microsoft.com;
 s=arcselector9901;
 h=From:Date:Subject:Message-ID:Content-Type:MIME-Version:X-MS-Exchange-SenderADCheck;
 bh=bXF8fTPJ/Y9pFcClOLMeQ5tenPBU4qylEmby8f9sbW8=;
 b=iZF7Z6ya9RARFt6c2+rWsAVvnHhbindfxsWuOA8KGAIb2uMy+dISq4uj7QafZBDa+kRD81oEOC997lM3FY3ZNJSYkYS88EiwjlV1Y2CBwsWhm4pz98dpIvu5WV13OfOUa3UymiN3/TKBAmrApkXj3i2ONfXwzRjOCn7SSdIMugPwbWEGDBDK0mBjjGc8uftpDCedkO9ukXB8Z4zJuJgdjpRgM0ajlgrua1V8xE1/HWu/jB69hWVJi1xYhP7oMnLcrjUNUlcMzIiVvBQFeBeV4SHSch2QVS7OzXDaJ21gIi+aoRw5GVOI/CBs8t0D4+4Teb9HaO3skhdfaxuNBwOxhA==
ARC-Authentication-Results: i=1; mx.microsoft.com 1; spf=pass
 smtp.mailfrom=amd.com; dmarc=pass action=none header.from=amd.com; dkim=pass
 header.d=amd.com; arc=none
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
 d=amdcloud.onmicrosoft.com; s=selector2-amdcloud-onmicrosoft-com;
 h=From:Date:Subject:Message-ID:Content-Type:MIME-Version:X-MS-Exchange-SenderADCheck;
 bh=bXF8fTPJ/Y9pFcClOLMeQ5tenPBU4qylEmby8f9sbW8=;
 b=lO8L1F44dYbQrV5PRYOwpt2OOEg9EvFfRa1pUHqqwt3blSFWJs3Gwnp9j4ExpUD6QIIYBduQdOurDxCio13ZnLyK1BX54JF8K94bGJTeOi8aCQWDS9NF4HsEarA3w247PIVFAW9yD33NZz2x/yuhHlAd0H8BpFJX0NF623fP1wI=
Authentication-Results: nongnu.org; dkim=none (message not signed)
 header.d=none;nongnu.org; dmarc=none action=none header.from=amd.com;
Received: from CY4PR12MB1352.namprd12.prod.outlook.com (2603:10b6:903:3a::13)
 by CY4PR12MB1607.namprd12.prod.outlook.com (2603:10b6:910:b::14) with
 Microsoft SMTP Server (version=TLS1_2,
 cipher=TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384) id 15.20.3370.16; Tue, 15 Sep
 2020 21:30:34 +0000
Received: from CY4PR12MB1352.namprd12.prod.outlook.com
 ([fe80::989b:b1b2:464c:443]) by CY4PR12MB1352.namprd12.prod.outlook.com
 ([fe80::989b:b1b2:464c:443%10]) with mapi id 15.20.3370.019; Tue, 15 Sep 2020
 21:30:33 +0000
From: Tom Lendacky <thomas.lendacky@amd.com>
To: qemu-devel@nongnu.org, kvm@vger.kernel.org
Cc: Marcel Apfelbaum <marcel.apfelbaum@gmail.com>,
        Paolo Bonzini <pbonzini@redhat.com>,
        "Dr. David Alan Gilbert" <dgilbert@redhat.com>,
        Eduardo Habkost <ehabkost@redhat.com>,
        Richard Henderson <rth@twiddle.net>,
        Connor Kuehl <ckuehl@redhat.com>,
        Brijesh Singh <brijesh.singh@amd.com>,
        Jiri Slaby <jslaby@suse.cz>,
        Marcelo Tosatti <mtosatti@redhat.com>,
        "Michael S. Tsirkin" <mst@redhat.com>
Subject: [PATCH v3 4/5] sev/i386: Don't allow a system reset under an SEV-ES
 guest
Date: Tue, 15 Sep 2020 16:29:43 -0500
Message-Id: 
 <058dcb33a9cc223e3180133d29e7a92bfdc40938.1600205384.git.thomas.lendacky@amd.com>
X-Mailer: git-send-email 2.28.0
In-Reply-To: <cover.1600205384.git.thomas.lendacky@amd.com>
References: <cover.1600205384.git.thomas.lendacky@amd.com>
X-ClientProxiedBy: DM6PR21CA0024.namprd21.prod.outlook.com
 (2603:10b6:5:174::34) To CY4PR12MB1352.namprd12.prod.outlook.com
 (2603:10b6:903:3a::13)
MIME-Version: 1.0
X-MS-Exchange-MessageSentRepresentingType: 1
Received: from tlendack-t1.amd.com (165.204.77.1) by
 DM6PR21CA0024.namprd21.prod.outlook.com (2603:10b6:5:174::34) with Microsoft
 SMTP Server (version=TLS1_2,
 cipher=TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384) id 15.20.3412.1 via Frontend
 Transport; Tue, 15 Sep 2020 21:30:32 +0000
X-Mailer: git-send-email 2.28.0
X-Originating-IP: [165.204.77.1]
X-MS-PublicTrafficType: Email
X-MS-Office365-Filtering-HT: Tenant
X-MS-Office365-Filtering-Correlation-Id: 400fbeb3-719d-495c-c1e2-08d859be9416
X-MS-TrafficTypeDiagnostic: CY4PR12MB1607:
X-MS-Exchange-Transport-Forked: True
X-Microsoft-Antispam-PRVS: 
 <CY4PR12MB160700615C797537AA830B9BEC200@CY4PR12MB1607.namprd12.prod.outlook.com>
X-MS-Oob-TLC-OOBClassifiers: OLM:418;
X-MS-Exchange-SenderADCheck: 1
X-Microsoft-Antispam: BCL:0;
X-Microsoft-Antispam-Message-Info: 
 wl4pLzrUN4iZXeFltD1AXlkTQIP2YIoKC3AlR2md2CKgbFj5DLDxc1fdWgkEZqsjxPQ0r/bd9yQgB0JK1yDi8Um2q1N7hlWlcOiEaptQoSaxNyoJ12DCMQDvpmcZmcgzuZSij+PfhLfLXX1IGAmjLJ3FEDkO/sXZZlvBKTIT7elJMsXVjTRKC6nueba9ORa2Sk0N1XlRvC7TPuwKgyZALRyimAkx0dkLjP4cM1GPQEH2vtzXqSxoBIMRjGdb4tvp45PScqRlFh/2JSB6pBras7ZwFBNBaizyI+UZc9HD5IF3KKIlczXds0TRKBU8G9CL
X-Forefront-Antispam-Report: 
 CIP:255.255.255.255;CTRY:;LANG:en;SCL:1;SRV:;IPV:NLI;SFV:NSPM;H:CY4PR12MB1352.namprd12.prod.outlook.com;PTR:;CAT:NONE;SFS:(4636009)(396003)(366004)(346002)(136003)(39860400002)(376002)(5660300002)(956004)(478600001)(6666004)(86362001)(52116002)(2616005)(2906002)(7696005)(36756003)(66476007)(8936002)(66556008)(66946007)(316002)(54906003)(7416002)(26005)(6486002)(4326008)(8676002)(16526019)(83380400001)(186003);DIR:OUT;SFP:1101;
X-MS-Exchange-AntiSpam-MessageData: 
 FcvrEI1uBWjZuelOaGS5d/PHx83Yo/QJwB+ZchJEm/cEab7qQ0m+mjp6JosobUENkRYbXIJdrwquCE+yPHeBTqxq19V6riVVorC9hOo+cJSh10mgANqCVB6+BEkvDijv9+q6XvJDmBGE1BN7jquFFHeV1r9imNwqof7/9k/kNIaTsEST0Km+WHKIxGqV58ElEVJIDykgf9/DuTXkKjTY9nZhn88UgTjBxxJABpl0IOFTCuow2GdFAdOAxSbpPViDCKsfaY/mQ8w0NKPAms/m+bD0lvYze+2HuT9pDmJtQkSNDDyemujW6wATs49hDQ/mCtxcDfKDa4UmRSSQN8VIEDdUyYWH0dLgkTN7Z6aO0gSgwV0yOGPuiYmBS2dJBs+2ZF2W8LzaglTs8Wyse1GCd1Ir+Ie4lwDS49NARIeYoqWRAtvo+qD6gMkCn0h4iXSQKzv7kcJdGnyUFIqjz+hNYA61MG6hLxMk73J6Vv+puFdhuuWGEI3tO1mD5lVlq9Xgd31AozgK2TZVQF11QKpmpGi6+gs6/fF1VShA/eYqMZD/mZ6x4DclorktScIauojO1paCdVWkG2NHpBQZVE5pJa6DriXbQzXkd9uRlfi5p/Em7XnZV1opqT5XPe0gxIcQF+lcW7NC0YCSyUztiQwDIA==
X-OriginatorOrg: amd.com
X-MS-Exchange-CrossTenant-Network-Message-Id: 
 400fbeb3-719d-495c-c1e2-08d859be9416
X-MS-Exchange-CrossTenant-AuthSource: CY4PR12MB1352.namprd12.prod.outlook.com
X-MS-Exchange-CrossTenant-AuthAs: Internal
X-MS-Exchange-CrossTenant-OriginalArrivalTime: 15 Sep 2020 21:30:33.8331
 (UTC)
X-MS-Exchange-CrossTenant-FromEntityHeader: Hosted
X-MS-Exchange-CrossTenant-Id: 3dd8961f-e488-4e60-8e11-a82d994e183d
X-MS-Exchange-CrossTenant-MailboxType: HOSTED
X-MS-Exchange-CrossTenant-UserPrincipalName: 
 OhCEAdmp/u4ySPQ7QKhb7acffs4LNH8mYN9mcjUU6GNGf9983RVbuwX3rhXhmY64JEe1UQTelEuc49+kpjmA6w==
X-MS-Exchange-Transport-CrossTenantHeadersStamped: CY4PR12MB1607
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org

From: Tom Lendacky <thomas.lendacky@amd.com>

An SEV-ES guest does not allow register state to be altered once it has
been measured. When a SEV-ES guest issues a reboot command, Qemu will
reset the vCPU state and resume the guest. This will cause failures under
SEV-ES, so prevent that from occurring.

Signed-off-by: Tom Lendacky <thomas.lendacky@amd.com>
---
 accel/kvm/kvm-all.c       | 9 +++++++++
 include/sysemu/cpus.h     | 2 ++
 include/sysemu/hw_accel.h | 5 +++++
 include/sysemu/kvm.h      | 2 ++
 softmmu/cpus.c            | 5 +++++
 softmmu/vl.c              | 5 ++++-
 6 files changed, 27 insertions(+), 1 deletion(-)

diff --git a/accel/kvm/kvm-all.c b/accel/kvm/kvm-all.c
index 20725b0368..63153b6e53 100644
--- a/accel/kvm/kvm-all.c
+++ b/accel/kvm/kvm-all.c
@@ -2388,6 +2388,15 @@ void kvm_flush_coalesced_mmio_buffer(void)
     s->coalesced_flush_in_progress = false;
 }
 
+bool kvm_cpu_check_resettable(void)
+{
+    /*
+     * If we have a valid reset vector override, then SEV-ES is active
+     * and the CPU can't be reset.
+     */
+    return !kvm_state->reset_valid;
+}
+
 static void do_kvm_cpu_synchronize_state(CPUState *cpu, run_on_cpu_data arg)
 {
     if (!cpu->vcpu_dirty) {
diff --git a/include/sysemu/cpus.h b/include/sysemu/cpus.h
index 3c1da6a018..6d688c757f 100644
--- a/include/sysemu/cpus.h
+++ b/include/sysemu/cpus.h
@@ -24,6 +24,8 @@ void dump_drift_info(void);
 void qemu_cpu_kick_self(void);
 void qemu_timer_notify_cb(void *opaque, QEMUClockType type);
 
+bool cpu_is_resettable(void);
+
 void cpu_synchronize_all_states(void);
 void cpu_synchronize_all_post_reset(void);
 void cpu_synchronize_all_post_init(void);
diff --git a/include/sysemu/hw_accel.h b/include/sysemu/hw_accel.h
index e128f8b06b..8b4536e7ae 100644
--- a/include/sysemu/hw_accel.h
+++ b/include/sysemu/hw_accel.h
@@ -17,6 +17,11 @@
 #include "sysemu/hvf.h"
 #include "sysemu/whpx.h"
 
+static inline bool cpu_check_resettable(void)
+{
+    return kvm_enabled() ? kvm_cpu_check_resettable() : true;
+}
+
 static inline void cpu_synchronize_state(CPUState *cpu)
 {
     if (kvm_enabled()) {
diff --git a/include/sysemu/kvm.h b/include/sysemu/kvm.h
index f74cfa85ab..eb94bbbff9 100644
--- a/include/sysemu/kvm.h
+++ b/include/sysemu/kvm.h
@@ -494,6 +494,8 @@ int kvm_physical_memory_addr_from_host(KVMState *s, void *ram_addr,
 
 #endif /* NEED_CPU_H */
 
+bool kvm_cpu_check_resettable(void);
+
 void kvm_cpu_synchronize_state(CPUState *cpu);
 void kvm_cpu_synchronize_post_reset(CPUState *cpu);
 void kvm_cpu_synchronize_post_init(CPUState *cpu);
diff --git a/softmmu/cpus.c b/softmmu/cpus.c
index a802e899ab..32f286643f 100644
--- a/softmmu/cpus.c
+++ b/softmmu/cpus.c
@@ -927,6 +927,11 @@ void hw_error(const char *fmt, ...)
     abort();
 }
 
+bool cpu_is_resettable(void)
+{
+    return cpu_check_resettable();
+}
+
 void cpu_synchronize_all_states(void)
 {
     CPUState *cpu;
diff --git a/softmmu/vl.c b/softmmu/vl.c
index 4eb9d1f7fd..422fbb1650 100644
--- a/softmmu/vl.c
+++ b/softmmu/vl.c
@@ -1475,7 +1475,10 @@ void qemu_system_guest_crashloaded(GuestPanicInformation *info)
 
 void qemu_system_reset_request(ShutdownCause reason)
 {
-    if (no_reboot && reason != SHUTDOWN_CAUSE_SUBSYSTEM_RESET) {
+    if (!cpu_is_resettable()) {
+        error_report("cpus are not resettable, terminating");
+        shutdown_requested = reason;
+    } else if (no_reboot && reason != SHUTDOWN_CAUSE_SUBSYSTEM_RESET) {
         shutdown_requested = reason;
     } else {
         reset_requested = reason;

From patchwork Tue Sep 15 21:29:44 2020
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Tom Lendacky <thomas.lendacky@amd.com>
X-Patchwork-Id: 11777877
Return-Path: <SRS0=V1hm=CY=vger.kernel.org=kvm-owner@kernel.org>
Received: from mail.kernel.org (pdx-korg-mail-1.web.codeaurora.org
 [172.30.200.123])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id 4B4F7139F
	for <patchwork-kvm@patchwork.kernel.org>;
 Tue, 15 Sep 2020 21:42:34 +0000 (UTC)
Received: from vger.kernel.org (vger.kernel.org [23.128.96.18])
	by mail.kernel.org (Postfix) with ESMTP id 24A1320809
	for <patchwork-kvm@patchwork.kernel.org>;
 Tue, 15 Sep 2020 21:42:34 +0000 (UTC)
Authentication-Results: mail.kernel.org;
	dkim=pass (1024-bit key) header.d=amdcloud.onmicrosoft.com
 header.i=@amdcloud.onmicrosoft.com header.b="A9sNDD63"
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1727928AbgIOVmY (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Tue, 15 Sep 2020 17:42:24 -0400
Received: from mail-dm6nam10on2050.outbound.protection.outlook.com
 ([40.107.93.50]:30689
        "EHLO NAM10-DM6-obe.outbound.protection.outlook.com"
        rhost-flags-OK-OK-OK-FAIL) by vger.kernel.org with ESMTP
        id S1728124AbgIOVbt (ORCPT <rfc822;kvm@vger.kernel.org>);
        Tue, 15 Sep 2020 17:31:49 -0400
ARC-Seal: i=1; a=rsa-sha256; s=arcselector9901; d=microsoft.com; cv=none;
 b=RvMRp8ykQ7ljWVSPTLhFhECDgwjC6IdNio7GrMQEiv/ATSP7flgZlUTBGP+wv7VmRmNzaoRyhlzIuJmu52g399n6QThCGYYwToXbmORBCoRKV6e1o6NoTimtKIIzHhjUTUP8xMI3tcW1nllgabZTjCA+FL7dFTzVMMLMZiJUzzsUw6iOf1PKkBaPz7NtglWrCu0fVlIBgDDf32EB8Erui8Bp6Qef8QpczHKDwS43xa8vTvvhRnV92CV9P31JmDLUW0tXFO7Xsus7lQIFqUbSf63AMADnXuZTv5XCbaTN6gt3CgtAAy89TA/rinEbaoDPJB0LFE14NUcIK7Fh1gDG0w==
ARC-Message-Signature: i=1; a=rsa-sha256; c=relaxed/relaxed; d=microsoft.com;
 s=arcselector9901;
 h=From:Date:Subject:Message-ID:Content-Type:MIME-Version:X-MS-Exchange-SenderADCheck;
 bh=5ycV0xdRknSZ3vAEGYSJeRPZJ+9Q3xBqw3WL1RRP48E=;
 b=E/pFPejiuKS8zgFGSjI00roR8l9MUUwO6UiWjC2q/+6HU/cGKjXH80lh/PxTBV8z8Q6PffcLihFsVbT4B3b0jnPfNrBJkDpn+t0C+cUCNqHsjLvImo6Qjl5CGahs/CmoDTxShoVFq7lmQ2AFa5dF/Ye2I71ivaH/iUSjHHdasfCurkBkWqlTd0KlDaopp+7RDgGo/RbItboq2QUvrqtm5kEaT0nlN3LKfoI1t2Gag52OIKuevcKylrbV32jKh20OZjQrnCwjofuMwyIddC1yTH/stmLTdtQJL2uph4SOdL/PCoMRJYrWK+S68BFYQYhc03Fr7zZMiyn4Jo9KNefF/w==
ARC-Authentication-Results: i=1; mx.microsoft.com 1; spf=pass
 smtp.mailfrom=amd.com; dmarc=pass action=none header.from=amd.com; dkim=pass
 header.d=amd.com; arc=none
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
 d=amdcloud.onmicrosoft.com; s=selector2-amdcloud-onmicrosoft-com;
 h=From:Date:Subject:Message-ID:Content-Type:MIME-Version:X-MS-Exchange-SenderADCheck;
 bh=5ycV0xdRknSZ3vAEGYSJeRPZJ+9Q3xBqw3WL1RRP48E=;
 b=A9sNDD63jtY2QQbZ5pSrjENuxq9GnIdyPe3WQ30nwc1AQr6LHjEk4Lk5YB7l/75YzhAg7jP8/XscJGHgqSjYV8+/LQ5PPrqtAhU7E3ANZq0R8z5gTxCtur5jhjHdF2D+n/SS/dMAcVhQaa/6Kpsu72E6j75FMQuVThoCvCDQCio=
Authentication-Results: nongnu.org; dkim=none (message not signed)
 header.d=none;nongnu.org; dmarc=none action=none header.from=amd.com;
Received: from CY4PR12MB1352.namprd12.prod.outlook.com (2603:10b6:903:3a::13)
 by CY4PR12MB1607.namprd12.prod.outlook.com (2603:10b6:910:b::14) with
 Microsoft SMTP Server (version=TLS1_2,
 cipher=TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384) id 15.20.3370.16; Tue, 15 Sep
 2020 21:30:42 +0000
Received: from CY4PR12MB1352.namprd12.prod.outlook.com
 ([fe80::989b:b1b2:464c:443]) by CY4PR12MB1352.namprd12.prod.outlook.com
 ([fe80::989b:b1b2:464c:443%10]) with mapi id 15.20.3370.019; Tue, 15 Sep 2020
 21:30:42 +0000
From: Tom Lendacky <thomas.lendacky@amd.com>
To: qemu-devel@nongnu.org, kvm@vger.kernel.org
Cc: Marcel Apfelbaum <marcel.apfelbaum@gmail.com>,
        Paolo Bonzini <pbonzini@redhat.com>,
        "Dr. David Alan Gilbert" <dgilbert@redhat.com>,
        Eduardo Habkost <ehabkost@redhat.com>,
        Richard Henderson <rth@twiddle.net>,
        Connor Kuehl <ckuehl@redhat.com>,
        Brijesh Singh <brijesh.singh@amd.com>,
        Jiri Slaby <jslaby@suse.cz>,
        Marcelo Tosatti <mtosatti@redhat.com>,
        "Michael S. Tsirkin" <mst@redhat.com>
Subject: [PATCH v3 5/5] sev/i386: Enable an SEV-ES guest based on SEV policy
Date: Tue, 15 Sep 2020 16:29:44 -0500
Message-Id: 
 <8e560a8577066c07b5bf1e5993fbd6d697702384.1600205384.git.thomas.lendacky@amd.com>
X-Mailer: git-send-email 2.28.0
In-Reply-To: <cover.1600205384.git.thomas.lendacky@amd.com>
References: <cover.1600205384.git.thomas.lendacky@amd.com>
X-ClientProxiedBy: DM6PR02CA0120.namprd02.prod.outlook.com
 (2603:10b6:5:1b4::22) To CY4PR12MB1352.namprd12.prod.outlook.com
 (2603:10b6:903:3a::13)
MIME-Version: 1.0
X-MS-Exchange-MessageSentRepresentingType: 1
Received: from tlendack-t1.amd.com (165.204.77.1) by
 DM6PR02CA0120.namprd02.prod.outlook.com (2603:10b6:5:1b4::22) with Microsoft
 SMTP Server (version=TLS1_2,
 cipher=TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384) id 15.20.3370.19 via Frontend
 Transport; Tue, 15 Sep 2020 21:30:41 +0000
X-Mailer: git-send-email 2.28.0
X-Originating-IP: [165.204.77.1]
X-MS-PublicTrafficType: Email
X-MS-Office365-Filtering-HT: Tenant
X-MS-Office365-Filtering-Correlation-Id: d0bdae57-f3b5-47ce-5db6-08d859be9951
X-MS-TrafficTypeDiagnostic: CY4PR12MB1607:
X-MS-Exchange-Transport-Forked: True
X-Microsoft-Antispam-PRVS: 
 <CY4PR12MB160713CA51DE6DE40B3F6A26EC200@CY4PR12MB1607.namprd12.prod.outlook.com>
X-MS-Oob-TLC-OOBClassifiers: OLM:9508;
X-MS-Exchange-SenderADCheck: 1
X-Microsoft-Antispam: BCL:0;
X-Microsoft-Antispam-Message-Info: 
 QphyhJbXiqa7nYb4sFtH+ax/7xNmay9vwA2q9Cu+P6Ox7FGuS7VMjOqRlg2MFDN17ZzeyQrhLomEV810IbTuCmsXuE66r2lYAqQX8dU5SF0uBMiNMocEy9IbFDZvbhN2GzeLt7a60PH7dc9BPwRgLArQyCdWVgj++KUqF5sfs8Ny6lMw+EJ4NGM7jADXnWhRL/r/ITrZ6xAgVV+V+4vLbeXue220t4Ami081Vxd9wH8JlZ4JF4yygbgsfpfknB0SrOIKQ7ACWu9u8pXhc332rqHrQJBKT2bIb+bxtKZAPv1PyYzav7pi8YvPPLkfedkpUU8x9+oQ+OYLEtp/GunnRg==
X-Forefront-Antispam-Report: 
 CIP:255.255.255.255;CTRY:;LANG:en;SCL:1;SRV:;IPV:NLI;SFV:NSPM;H:CY4PR12MB1352.namprd12.prod.outlook.com;PTR:;CAT:NONE;SFS:(4636009)(396003)(366004)(346002)(136003)(39860400002)(376002)(5660300002)(956004)(478600001)(86362001)(52116002)(2616005)(2906002)(7696005)(36756003)(66476007)(8936002)(66556008)(66946007)(316002)(54906003)(7416002)(26005)(6486002)(4326008)(8676002)(16526019)(83380400001)(186003);DIR:OUT;SFP:1101;
X-MS-Exchange-AntiSpam-MessageData: 
 47GMNjAWOuJCdlJy2GSLH/1ruByqDPly69GECDd9SQUAnLyIPLv/tTCK7RHsJc8tx/vG+Z4mEDX5NujLTEMQPwSInB8CbOn3A2Aq1IyE62mMuOyHSwo8/GTAqKLzn7jkCf5wNionEB0AIKhqTvEIV2n9CpJ2pOUucRSz2X9Xazm/DJdhuag4r1vT2OPMAJxSRBErMhWdtyFharkr3sJtckYR72Mvghm2AehNrKAn/omKjYsJnjI1CYCRYGR6LqSlARBcfHk4RJkaQc0ahnclbEdAprhofWBqMlQUXPb6IAoP4BnoaMTHtKfpWn0KsNhqqi5pR5Zv8vbRyJip1psr+P0SW+AXq8Zxgv0Y/suUxiXBXwLDRK/GtiyIh3250IlJdzq+2T+Co2Swt0wFzMtK+t/Ez3Q5AKHHiQ3xKdE08+DzI9snfHf5+krIw1ihcfy8ne5m5boujGNlFCMVvisg6xSHa71tsuKp/bmVQqZXW6kGexT0XDixyTrULVls9M2JXSsEuQISofNuvgxQ5ml+UNSqyHhp22Udouf9+6jaynbWNo/sq59ugxozsisMjJMn3SgU3G7GHHG8yeTckWl1FGjmwG0Xer/ij57N4c8miYRaVSZEfyxOiYH5mx5OFJn3yGTWbtDTnd0NZwuXQu0LRg==
X-OriginatorOrg: amd.com
X-MS-Exchange-CrossTenant-Network-Message-Id: 
 d0bdae57-f3b5-47ce-5db6-08d859be9951
X-MS-Exchange-CrossTenant-AuthSource: CY4PR12MB1352.namprd12.prod.outlook.com
X-MS-Exchange-CrossTenant-AuthAs: Internal
X-MS-Exchange-CrossTenant-OriginalArrivalTime: 15 Sep 2020 21:30:42.6028
 (UTC)
X-MS-Exchange-CrossTenant-FromEntityHeader: Hosted
X-MS-Exchange-CrossTenant-Id: 3dd8961f-e488-4e60-8e11-a82d994e183d
X-MS-Exchange-CrossTenant-MailboxType: HOSTED
X-MS-Exchange-CrossTenant-UserPrincipalName: 
 2JkiRUzGHCOrSK5dh8YCB6V0rjvT/6VOCHfdhWz2VWvv6zS55QbDBor+oa1s0JyR0dChXKzMIxbQTA62+ReL1g==
X-MS-Exchange-Transport-CrossTenantHeadersStamped: CY4PR12MB1607
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org

From: Tom Lendacky <thomas.lendacky@amd.com>

Update the sev_es_enabled() function return value to be based on the SEV
policy that has been specified. SEV-ES is enabled if SEV is enabled and
the SEV-ES policy bit is set in the policy object.

Signed-off-by: Tom Lendacky <thomas.lendacky@amd.com>
---
 target/i386/sev.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/target/i386/sev.c b/target/i386/sev.c
index 6ddefc65fa..bcaadaa2f9 100644
--- a/target/i386/sev.c
+++ b/target/i386/sev.c
@@ -70,6 +70,8 @@ struct SevGuestState {
 #define DEFAULT_GUEST_POLICY    0x1 /* disable debug */
 #define DEFAULT_SEV_DEVICE      "/dev/sev"
 
+#define GUEST_POLICY_SEV_ES_BIT (1 << 2)
+
 /* SEV Information Block GUID = 00f771de-1a7e-4fcb-890e-68c77e2fb44e */
 #define SEV_INFO_BLOCK_GUID \
     "\xde\x71\xf7\x00\x7e\x1a\xcb\x4f\x89\x0e\x68\xc7\x7e\x2f\xb4\x4e"
@@ -375,7 +377,7 @@ sev_enabled(void)
 bool
 sev_es_enabled(void)
 {
-    return false;
+    return sev_enabled() && (sev_guest->policy & GUEST_POLICY_SEV_ES_BIT);
 }
 
 uint64_t
